src/sankey.js

```diff
@@ -0,0 +1,152 @@
+import React, {PropTypes, Component, cloneElement} from 'react';
+import shallowCompare from 'react-addons-shallow-compare';
+import linkKey from './utils/link-key';
+import getWidthSums from './utils/get-width-sums';
+import separateChildrenByType from './utils/separate-children-by-type';
+import findColumns from './utils/find-columns';
+import configureNodeCoordinates from './utils/configure-node-coordinates';
+import getAnchorFromRectangleNodes from './utils/get-anchor-from-rectangle-nodes';
+
+const {number} = PropTypes;
+export default class Sankey extends Component {
+
+  static propTypes = {
+    /** width of svg figure */
+    width: number.isRequired,
+    /** height of svg figure */
+    height: number.isRequired,
+    /** spacing (horizontal) */
+    spacing: number,
+    /** margin (vertical) */
+    margin: number
+  };
+
+  shouldComponentUpdate(nextProp, nextState) {
+    return shallowCompare(this, nextProp, nextState);
+  }
+
+  render() {
+    const {
+      children,
+      width: containerWidth,
+      height: containerHeight,
+      spacing,
+      margin,
+      ..._props
+    } = this.props;
+
+    const {defs, nodes, links} = separateChildrenByType(children);
+
+    const columns = findColumns(nodes, links);
+
+    const defaultColumnWidth = (containerWidth - columns.length * spacing) /
+      (columns.length + 1);
+
+
+    /* if the max width of the column is less than the default column with,
+     * take the bigger value. When no width is given, a default column width
+     * is used. */
+    const columnWidths = columns.map(
+      column =>
+        (Math.max.apply(null, column.map(
+          ({props: {width, r}}) => width || r * 2)
+        ) || defaultColumnWidth)
+    );
+
+    const nodesWithCoords =
+      configureNodeCoordinates(
+        columns,
+        nodes,
+        links,
+        columnWidths,
+        spacing,
+        margin
+      );
+
+    const nodeYs = nodesWithCoords
+      .reduce((hash, {props: {name, y, height}}) => {
+        return {
+          ...hash,
+          [name]: (y + height / 2)
+        };
+      }, {});
+
+    // order links by the from and to block vertical position (Y).
+    // this is the layout logic that prevents connectors from crossing
+    // each other.
+    const orderedLinks = links.sort(
+      ({props: {to: to1, from: from1}},
+        {props: {to: to2, from: from2}}) => (
+        nodeYs[to1] - nodeYs[to2] +
+        nodeYs[from1] - nodeYs[from2]
+      )
+    );
+
+    const linkWidths = links
+      .reduce((hash, {props: {from, to, width}}) => {
+        return {
+          ...hash,
+          [linkKey(from, to)]: width
```

Comment:
Is linkKey used in node-link graphs? if yes, what will happen if there are two connections from one node to another?